src/main/java/org/opensearch/security/support/Base64JDKHelper.java

```diff
@@ -0,0 +1,199 @@
+            try {
+                return useSafeObjectOutputStream ? new SafeObjectOutputStream(out) : new ObjectOutputStream(out);
+            } catch (SecurityException e) {
+                // As we try to create SafeObjectOutputStream only when necessary permissions are granted, we should
+                // not reach here, but if we do, we can still return ObjectOutputStream after resetting ByteArrayOutputStream
+                out.reset();
+                return new ObjectOutputStream(out);
+            }
+        }
+
+        @SuppressWarnings("removal")
+        private SafeObjectOutputStream(OutputStream out) throws IOException {
+            super(out);
+
+            SecurityManager sm = System.getSecurityManager();
+            if (sm != null) {
+                sm.checkPermission(new SpecialPermission());
+            }
+
+            AccessController.doPrivileged((PrivilegedAction<Boolean>) () -> enableReplaceObject(true));
+        }
+
+        @Override
+        protected Object replaceObject(Object obj) throws IOException {
+            Class<?> clazz = obj.getClass();
+            if (isSafeClass(clazz)) {
+                return obj;
+            }
+            throw new IOException("Unauthorized serialization attempt " + clazz.getName());
+        }
+    }
+
+    public static String serializeObject(final Serializable object) {
+
+        Preconditions.checkArgument(object != null, "object must not be null");
+
+        final ByteArrayOutputStream bos = new ByteArrayOutputStream();
+        try (final ObjectOutputStream out = SafeObjectOutputStream.create(bos)) {
+            out.writeObject(object);
+        } catch (final Exception e) {
+            throw new OpenSearchException("Instance {} of class {} is not serializable", e, object, object.getClass());
+        }
+        final byte[] bytes = bos.toByteArray();
+        return BaseEncoding.base64().encode(bytes);
+    }
+
+    public static Serializable deserializeObject(final String string) {
+
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(string), "string must not be null or empty");
+
+        final byte[] bytes = BaseEncoding.base64().decode(string);
+        final ByteArrayInputStream bis = new ByteArrayInputStream(bytes);
+        try (SafeObjectInputStream in = new SafeObjectInputStream(bis)) {
+            return (Serializable) in.readObject();
+        } catch (final Exception e) {
+            throw new OpenSearchException(e);
+        }
+    }
+
+    private final static class SafeObjectInputStream extends ObjectInputStream {
+
+        public SafeObjectInputStream(InputStream in) throws IOException {
+            super(in);
+        }
+
+        @Override
+        protected Class<?> resolveClass(ObjectStreamClass desc) throws IOException, ClassNotFoundException {
+
+            Class<?> clazz = super.resolveClass(desc);
+            if (isSafeClass(clazz)) {
```

Comment:
Don't we need similar checks with custom serialization as well?

Reply:
Replied here - https://github.com/opensearch-project/security/pull/2802#discussion_r1257688398